presto-main/src/main/java/io/prestosql/type/DateTimes.java

```diff
@@ -474,6 +494,57 @@ private static long toEpochSecond(String year, String month, String day, String
         return timestamp.toEpochSecond(offsets.get(0));
     }
 
+    public static boolean timeHasTimeZone(String value)
+    {
+        Matcher matcher = TIME_PATTERN.matcher(value);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException(format("Invalid time '%s'", value));
+        }
+
+        return matcher.group("timezone") != null;
+    }
+
+    public static int extractTimePrecision(String value)
+    {
+        Matcher matcher = TIME_PATTERN.matcher(value);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException(format("Invalid time '%s'", value));
+        }
+
+        String fraction = matcher.group("fraction");
+        if (fraction == null) {
+            return 0;
+        }
+
+        return fraction.length();
+    }
+
+    public static long parseTime(String value)
+    {
+        Matcher matcher = TIME_PATTERN.matcher(value);
+        if (!matcher.matches() || matcher.group("timezone") != null) {
+            throw new IllegalArgumentException("Invalid time: " + value);
+        }
+
+        int hour = Integer.parseInt(matcher.group("hour"));
+        int minute = Integer.parseInt(matcher.group("minute"));
+        int second = matcher.group("second") == null ? 0 : Integer.parseInt(matcher.group("second"));
+
+        if (hour > 23 || minute > 59 || second > 59) {
+            throw new IllegalArgumentException("Invalid time: " + value);
+        }
+
+        int precision = 0;
+        String fraction = matcher.group("fraction");
+        long fractionValue = 0;
+        if (fraction != null) {
+            precision = fraction.length();
+            fractionValue = Long.parseLong(fraction);
+        }
+
+        return (((hour * 60) + minute) * 60 + second) * PICOSECONDS_PER_SECOND + rescale(fractionValue, precision, 12);
```

Comment:
should we limit fraction value to 12 digits?

Reply:
Yes, great point. I'll update and add tests.